projects/destinations_similarity/destinations_similarity/scraper/extractor.py

```diff
@@ -0,0 +1,164 @@
+                Defaults to ['p', 'li'].
+
+        Returns:
+            List[str]: A list with each piece of text extracted from the
+                specified tags.
+        """
+        tags = tags or ['p', 'li']
+        soup = BeautifulSoup(text, "html.parser")
+        return [
+            re.sub(r'\[.*?\]|<.*?>', '', str(x)).strip()
+            for x in soup.find_all(tags)
+        ]
+
+    def extract_images(self, page: str) -> List[str]:
+        """Retrieve images (as links) for a specified page.
+
+        Args:
+            page (str): The name of the page.
+
+        Returns:
+            List[str]: A list with the URLs of the images.
+        """
+        request = self.session.get(f"{self.rest_url}/page/media-list/{page}")
+        response = json.loads(request.text)
+        items = response.get('items', [])
+
+        images_links = []
+
+        for item in items:
+            if item['type'] == 'image' and 'srcset' in item:
+                images_links += [f"https:{item['srcset'][0]['src']}"]
+
+        return images_links
+
+    def extract_content_raw(
+        self, page: str, summary: bool, sections: List[str] = None
+    ) -> Dict[str, str]:
+        """Retrieve the HTML-formatted sections from a page.
+
+        Args:
+            page (str): The name of the page.
+            summary (bool): Boolean that specifies if the summary for the page
+                must be retrieved.
+            sections (List[str], optional): A list of sections to be retrieved.
+                Defaults to None.
+
+        Returns:
+            Dict[str, str]: A dictionary with the sections, where each key is
+                the section name.
+        """
+        sections = sections or []
+
+        request = self.session.get(
+            f"{self.rest_url}/page/mobile-sections/{page}")
+        response = json.loads(request.text)
+
+        sections_data = {}
+
+        # Retrieve summary
+        if summary and 'lead' in response:
+            sections_data['summary'] = response['lead']['sections'][0]['text']
+
+        # Retrieve sections and subsections (with HTML tags)
+        if sections and 'remaining' in response:
+            page_sections = response['remaining']['sections']
+
+            # Get index of sections found
+            idx_sections_found = [
+                i for i, section in enumerate(page_sections)
+                if section.get('line') in sections
+            ]
+
+            # Get level of each section, to identify subsections
+            levels = [section.get('toclevel', -2) for section in page_sections]
+
+            for start in idx_sections_found:
+                try:
+                    # Get index next section at same toclevel
+                    end = next(
+                        i + (start + 1)
+                        for i, level in enumerate(levels[start + 1:])
+                        if level <= levels[start]
+                    )
```

Comment:
This block of code pulls the first element of the inner list if I'm not wrong. So, would it be easy to implement this without the looping cause looping isn't required if only the first element is needed?

Reply:
This block of code (the entire section retrieval, actually) is more complex than it looks.

Every section has a level, and the higher the level, the more nested the section is. This means that a level 2 section is a subsection of a level 1 section.

The API gives us a flat list of every section on a page. To retrieve a section and its subsections, we have to look for the next instance of the same level (or higher) section. For example, consider that the sections and levels were:

```python
sections = [('History', 1), ('Independence', 2), ('War', 2), ('Geography', 1), ...]
```

If we want to retrieve the "History" section, we must also get the "Independence" and "War" subsections, so we have to slice the list into `sections[0:3]`. Retrieving the end index is what the `end = next(...)` block of code is doing: it's iterating through the list searching for the next element that delimits a new section!